airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py

```diff
@@ -0,0 +1,233 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+"""The actions package is made to declare reusable pipeline components."""
+
+from __future__ import annotations
+from dataclasses import dataclass, field
+import inspect
+from pipelines import main_logger
+
+from typing import TYPE_CHECKING, Callable, Dict, List, Tuple, Union
+
+import asyncer
+from pipelines.models.steps import StepStatus, StepResult
+
+RESULTS_DICT = Dict[str, StepResult]
+ARGS_TYPE = Union[Dict, Callable[[RESULTS_DICT], Dict]]
+
+if TYPE_CHECKING:
+    from pipelines.models.steps import Step, StepResult
+
+
+@dataclass
+class RunStepOptions:
+    """Options for the run_step function."""
+    fail_fast: bool = True
+    skip_steps: List[str] = field(default_factory=list)
+    log_step_tree: bool = True
+
+
+@dataclass(frozen=True)
+class StepToRun:
+    """
+    A class to wrap a Step with its id and args.
+
+    Used to coordinate the execution of multiple steps inside a pipeline.
+    """
+    id: str
+    step: Step
+    args: ARGS_TYPE = field(default_factory=dict)
+    depends_on: List[str] = field(default_factory=list)
+
+STEP_TREE = List[StepToRun | List[StepToRun]]
+
+
+async def evaluate_run_args(args: ARGS_TYPE, results: RESULTS_DICT) -> Dict:
+    """
+    Evaluate the args of a StepToRun using the results of previous steps.
+    """
+    if inspect.iscoroutinefunction(args):
+        return await args(results)
+    elif callable(args):
+        return args(results)
```

Comment:
According to your type hints `args` can be a `dict` but here you consider its always a `callable`

Reply:
Hmm maybe I dont follow but I think the type is correct 🤔 

```
ARGS_TYPE = Union[Dict, Callable[[RESULTS_DICT], Dict]]
```

Means args is either a `dict` or a function that returns a `dict`